### ISSUE TITLE
[BUG] JAX JIT-ing a QNode yields different results

### ISSUE THREAD

**Issue by erick-xanadu** [PennyLaneAI/pennylane#4734]
### Expected behavior

JAX JIT-ing a QNode yields the same results

### Actual behavior

Different results

### Additional information

I haven't reduced the code too much but this was the test case that triggered it. I run it with pytest.

### Source code

```python
import jax
import numpy as np
import pennylane as qml
import pytest

jax.config.update("jax_enable_x64", True)

class RX_broadcasted(qml.RX):
    """A version of qml.RX that detects batching."""

    ndim_params = (0,)
    compute_decomposition = staticmethod(lambda theta, wires=None: [qml.RX(theta, wires=wires)])


class TestBroadcastExpand:

    def test_expansion_qnode(self):
        param = np.array([0.1, 0.8, 2.1])
        size = 3

        def qnode_builder(device_name):
            @qml.transforms.broadcast_expand
            @qml.qnode(qml.device(device_name, wires=2), interface="jax")
            def circuit(y):
                RX_broadcasted(y, wires=1)
                return qml.state()
            return circuit

        qnode_lightning = qnode_builder("lightning.qubit")
        qnode_default = qnode_builder("default.qubit")

        expected = qnode_lightning(param)
        observed = qnode_default(param)
        assert np.allclose(expected, observed)

        expected = jax.jit(qnode_lightning)(param)
        observed = jax.jit(qnode_default)(param)

        assert np.allclose(expected, observed)
```

### Tracebacks

```shell
bug.py F                                                                                                                                                                                            [100%]

================================================================================================ FAILURES =================================================================================================
________________________________________________________________________________ TestBroadcastExpand.test_expansion_qnode _________________________________________________________________________________

self = <bug.TestBroadcastExpand object at 0x7fdea0842260>

    def test_expansion_qnode(self):
        param = np.array([0.1, 0.8, 2.1])
        size = 3
    
        def qnode_builder(device_name):
            @qml.transforms.broadcast_expand
            @qml.qnode(qml.device(device_name, wires=2), interface="jax")
            def circuit(y):
                RX_broadcasted(y, wires=1)
                return qml.state()
            return circuit
    
        qnode_lightning = qnode_builder("lightning.qubit")
        qnode_default = qnode_builder("default.qubit")
    
        expected = qnode_lightning(param)
        observed = qnode_default(param)
        assert np.allclose(expected, observed)
    
        expected = jax.jit(qnode_lightning)(param)
        observed = jax.jit(qnode_default)(param)
    
>       assert np.allclose(expected, observed)
E       assert False
E        +  where False = <function allclose at 0x7fdebe3edb30>(Array([[0.99875026+0.j        , 0.        -0.04997917j,\n        0.        +0.j        , 0.        +0.j        ],\n     ...5+0.j        , 0.        -0.86742323j,\n        0.        +0.j        , 0.        +0.j        ]],      dtype=complex128), Array([[0.99875026+0.j        , 0.        -0.04997917j,\n        0.        +0.j        , 0.        +0.j        ],\n     ...6+0.j        , 0.        -0.04997917j,\n        0.        +0.j        , 0.        +0.j        ]],      dtype=complex128))
E        +    where <function allclose at 0x7fdebe3edb30> = np.allclose

bug.py:39: AssertionError
-------------------------------------------------------------------------------------------- Captured log call --------------------------------------------------------------------------------------------
WARNING  jax._src.xla_bridge:xla_bridge.py:636 No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
========================================================================================= short test summary info =========================================================================================
FAILED bug.py::TestBroadcastExpand::test_expansion_qnode - assert False

```

### System information

```shell
(env) erick.ochoalopez@DL7420-GS4N1J3:~/Code/cataliist$ python
Python 3.10.11 (main, May 13 2023, 12:07:51) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pennylane as qml
>>> qml.about()
Name: PennyLane
Version: 0.34.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/erick.ochoalopez/Code/env/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: pennylane-catalyst, PennyLane-Lightning, PennyLane-Lightning-Kokkos

Platform info:           Linux-5.15.0-84-generic-x86_64-with-glibc2.31
Python version:          3.10.11
Numpy version:           1.26.1
Scipy version:           1.11.3
Installed devices:
- default.gaussian (PennyLane-0.34.0.dev0)
- default.mixed (PennyLane-0.34.0.dev0)
- default.qubit (PennyLane-0.34.0.dev0)
- default.qubit.autograd (PennyLane-0.34.0.dev0)
- default.qubit.jax (PennyLane-0.34.0.dev0)
- default.qubit.legacy (PennyLane-0.34.0.dev0)
- default.qubit.tf (PennyLane-0.34.0.dev0)
- default.qubit.torch (PennyLane-0.34.0.dev0)
- default.qutrit (PennyLane-0.34.0.dev0)
- null.qubit (PennyLane-0.34.0.dev0)
- lightning.kokkos (PennyLane-Lightning-Kokkos-0.32.0)
- lightning.qubit (PennyLane-Lightning-0.32.0)
>>>
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by erick-xanadu** [comment]
Maybe I am missing something but I realize that I had a comment that says that "Broadcast expand appears to only work on default.qubit." Am I doing something wrong in this code / breaking some assumptions?

**Comment by albi3ro** [comment]
Think you could create a more minimal failing example?

**Comment by erick-xanadu** [comment]
I got these examples from PL transform tests. I will try to minimize it.

**Comment by erick-xanadu** [comment]
@albi3ro: Here is the minimization

```python
import jax
import numpy as np
import pennylane as qml
import pytest

jax.config.update("jax_enable_x64", True)

class RX_broadcasted(qml.RX):
    """A version of qml.RX that detects batching."""

    ndim_params = (0,)
    compute_decomposition = staticmethod(lambda theta, wires=None: [qml.RX(theta, wires=wires)])


class TestBroadcastExpand:

    def test_expansion_qnode(self):
        param = np.array([0.1, 0.8, 2.1])
        size = 3

        def qnode_builder(device_name):
            @qml.transforms.broadcast_expand
            @qml.qnode(qml.device(device_name, wires=2), interface="jax")
            def circuit(y):
                RX_broadcasted(y, wires=1)
                return qml.state()
            return circuit

        qnode_lightning = qnode_builder("lightning.qubit")
        qnode_default = qnode_builder("default.qubit")

        expected = qnode_lightning(param)
        observed = qnode_default(param)
        assert np.allclose(expected, observed)

        expected = jax.jit(qnode_lightning)(param)
        observed = jax.jit(qnode_default)(param)

        assert np.allclose(expected, observed)
```

**Comment by erick-xanadu** [comment]
We don't need the custom `RX_broadcasted`

```python
import jax
import numpy as np
import pennylane as qml
import pytest

jax.config.update("jax_enable_x64", True)

class TestBroadcastExpand:

    def test_expansion_qnode(self):
        param = np.array([0.1, 0.8, 2.1])

        def qnode_builder(device_name):
            @qml.transforms.broadcast_expand
            @qml.qnode(qml.device(device_name, wires=2), interface="jax")
            def circuit(y):
                qml.RX(y, wires=1)
                return qml.state()
            return circuit

        qnode_lightning = qnode_builder("lightning.qubit")
        qnode_default = qnode_builder("default.qubit")

        expected = qnode_lightning(param)
        observed = qnode_default(param)
        assert np.allclose(expected, observed)

        expected = jax.jit(qnode_lightning)(param)
        observed = jax.jit(qnode_default)(param)
        print(expected)
        print(observed)

        assert np.allclose(expected, observed)
```

What is interesting is that the code does not fail the assertion if `@qml.transforms.broadcast_expand` is not used.

**Comment by albi3ro** [comment]
The offending problem is caching with backprop:

```
@qml.transforms.broadcast_expand
@qml.qnode(qml.device('default.qubit', wires=2), interface="jax", diff_method="backprop")
def circuit_backprop(y):
    qml.RX(y, wires=1)
    return qml.state()
```

This circuit gives different answers depending on whether its its broadcasted. If `cache=False`, then jitting gives the correct results.

By calling the batch transform, the tape is split up into multiple tapes that all have the same identical hash. The result for that first tape is then reused for all three tapes.

As really quick patch, we can potentially add a warning if jax-jit is used with backprop and a cache.

**Comment by albi3ro** [comment]
This is a duplicate of Issue #2762

**Comment by erick-xanadu** [comment]
Ah, you can close it as a duplicate if you want then and I can track the other ticket. Thanks @albi3ro.
